### PR TITLE
Add --log-level to ros2 bag play and record

### PR DESCRIPTION
**Pull request by r7vme**
- Depends from https://github.com/ros2/rosbag2/pull/1632
- 
- This change adds new option `--log-level` to `ros2 bag play` to allow print e.g. debug logs. Currently user can not change default logging level (info).

Examples:

info
```
ros2 bag play rosbag2_2024_04_23-12_01_27/ --log-level info
[INFO] [1713967388.214619740] [rosbag2_player]: Set rate to 1
[INFO] [1713967388.383060294] [rosbag2_player]: Adding keyboard callbacks.
[INFO] [1713967388.383322290] [rosbag2_player]: Press SPACE for Pause/Resume
[INFO] [1713967388.383947615] [rosbag2_player]: Press CURSOR_RIGHT for Play Next Message
[INFO] [1713967388.384285062] [rosbag2_player]: Press CURSOR_UP for Increase Rate 10%
[INFO] [1713967388.384607737] [rosbag2_player]: Press CURSOR_DOWN for Decrease Rate 10%
[INFO] [1713967388.386311806] [rosbag2_player]: Playback until timestamp: -1
[INFO] [1713967390.580246159] [rosbag2_player]: Stopping playback.
[INFO] [1713967390.755944560] [rclcpp]: signal_handler(signum=2)
```

debug
```
ros2 bag play rosbag2_2024_04_23-12_01_27/ --log-level debug
[DEBUG] [1713967406.290722714] [rclcpp]: signal handler installed
[DEBUG] [1713967406.290799689] [rclcpp]: deferred_signal_handler(): waiting for SIGINT/SIGTERM or uninstall
[DEBUG] [1713967406.298048672] [pluginlib.ClassLoader]: Creating ClassLoader, base = rosbag2_storage::storage_interfaces::ReadWriteInterface, address = 0x63cf452e4340
[DEBUG] [1713967406.300786500] [pluginlib.ClassLoader]: Entering determineAvailableClasses()...
[DEBUG] [1713967406.300803318] [pluginlib.ClassLoader]: Processing xml file /home/r7vme/ws/install/rosbag2_storage/share/rosbag2_storage/test/rosbag2_storage/test_plugin.xml...
[DEBUG] [1713967406.300891790] [pluginlib.ClassLoader]: XML file specifies lookup name (i.e. magic name) = my_test_plugin.
[DEBUG] [1713967406.300907280] [pluginlib.ClassLoader]: XML file specifies lookup name (i.e. magic name) = my_read_only_test_plugin.
[DEBUG] [1713967406.300913918] [pluginlib.ClassLoader]: Processing xml file /home/r7vme/ws/install/rosbag2_storage_mcap/share/rosbag2_storage_mcap/plugin_description.xml...
[DEBUG] [1713967406.300956120] [pluginlib.ClassLoader]: XML file specifies lookup name (i.e. magic name) = mcap.
[DEBUG] [1713967406.300966477] [pluginlib.ClassLoader]: Processing xml file /home/r7vme/ws/install/rosbag2_storage_sqlite3/share/rosbag2_storage_sqlite3/plugin_description.xml...
[DEBUG] [1713967406.301001733] [pluginlib.ClassLoader]: XML file specifies lookup name (i.e. magic name) = sqlite3.
[DEBUG] [1713967406.301013964] [pluginlib.ClassLoader]: Exiting determineAvailableClasses()...
[DEBUG] [1713967406.301021444] [pluginlib.ClassLoader]: Finished constructring ClassLoader, base = rosbag2_storage::storage_interfaces::ReadWriteInterface, address = 0x63cf452e4340
[DEBUG] [1713967406.301036392] [pluginlib.ClassLoader]: Creating ClassLoader, base = rosbag2_storage::storage_interfaces::ReadOnlyInterface, address = 0x63cf4563a9b0
[DEBUG] [1713967406.303476865] [pluginlib.ClassLoader]: Entering determineAvailableClasses()...
[DEBUG] [1713967406.303487897] [pluginlib.ClassLoader]: Processing xml file /home/r7vme/ws/install/rosbag2_storage/share/rosbag2_storage/test/rosbag2_storage/test_plugin.xml...
[DEBUG] [1713967406.303559663] [pluginlib.ClassLoader]: XML file specifies lookup name (i.e. magic name) = my_test_plugin.
[DEBUG] [1713967406.303567099] [pluginlib.ClassLoader]: XML file specifies lookup name (i.e. magic name) = my_read_only_test_plugin.
[DEBUG] [1713967406.303577954] [pluginlib.ClassLoader]: Processing xml file /home/r7vme/ws/install/rosbag2_storage_mcap/share/rosbag2_storage_mcap/plugin_description.xml...
[DEBUG] [1713967406.303631795] [pluginlib.ClassLoader]: XML file specifies lookup name (i.e. magic name) = mcap.
...skipped...
[DEBUG] [1713967407.670974917] [rcl]: Finalizing node
[DEBUG] [1713967407.674591216] [rcl]: Node finalized
[INFO] [1713967407.674653174] [rclcpp]: signal_handler(signum=2)
[DEBUG] [1713967407.674662064] [rclcpp]: signal_handler(): notifying deferred signal handler
[DEBUG] [1713967407.674759349] [rclcpp]: deferred_signal_handler(): woken up due to SIGINT/SIGTERM or uninstall
[DEBUG] [1713967407.674777948] [rclcpp]: deferred_signal_handler(): shutting down
[DEBUG] [1713967407.674771623] [rcl]: Shutting down ROS client library, for context at address: 0x63cf45598ec0
[DEBUG] [1713967407.674822040] [rclcpp]: deferred_signal_handler(): shutting down rclcpp::Context @ 0x63cf455a4560, because it had shutdown_on_signal == true
[DEBUG] [1713967407.674864531] [rclcpp]: SignalHandler::uninstall(): notifying deferred signal handler
[DEBUG] [1713967407.674874613] [rclcpp]: deferred_signal_handler(): signal handling uninstalled
[DEBUG] [1713967407.674913747] [rclcpp]: signal handler uninstalled
```

### CODE REVIEW

**Comment by r7vme** [comment]
@MichaelOrlov could you please take a look? this is draft PR to validate design. If design is OK, I'll add tests and also cover recorder

File: rosbag2_py/src/rosbag2_py/_transport.cpp
Comment (MichaelOrlov) [inline rosbag2_py/src/rosbag2_py/_transport.cpp]:
```suggestion
explicit Player(const std::string & log_level)
```
Use `explicit` qualifier to avoid occasional type conversion.

**Comment by MichaelOrlov** [comment]
@r7vme Thanks for the PR! The approach looks good to me. :+1:

File: ros2bag/ros2bag/verb/play.py
Comment (fujitatomoya) [inline ros2bag/ros2bag/verb/play.py]: How about moving this into the method like `add_loglevel_argument(parser)` to `api/__init__.py`?

```python
def add_loglevel_argument(parser):
    parser.add_argument(
       '--log-level', type=str, default='info',
       choices=['debug', 'info', 'warn', 'error', 'fatal'],
       help='Logging level'
    )
```

and we can call it here,

```suggestion
        add_loglevel_argument(parser)
```

so that this can be imported and used on else where, possibly on `Recorder` as well?

File: rosbag2_py/src/rosbag2_py/_transport.cpp
Comment (fujitatomoya) [inline rosbag2_py/src/rosbag2_py/_transport.cpp]: Does it make more sense to move this into `rosbag2_cpp`, it already has dependency on `rosbag2_cpp`?

**Review by fujitatomoya — APPROVED** [review]
lgtm with a couple of minor refactoring ideas.

File: ros2bag/ros2bag/verb/play.py
Comment (MichaelOrlov) [reply ros2bag/ros2bag/verb/play.py]: @fujitatomoya I would prefer to have a parameter's definition in one place for the sake of transparency. 
If one needs to look at or check what parameters are defined or exist, if it will be a call to the `add_loglevel_argument(parser)` will need to have an extra headache and "archeology search" to figure out where it is defined to see what is inside of it.
Duplication of the same parameter definitions for player and recorder is worth it here for clarity.

File: rosbag2_py/src/rosbag2_py/_transport.cpp
Comment (MichaelOrlov) [reply rosbag2_py/src/rosbag2_py/_transport.cpp]: @fujitatomoya As regards
> Does it make more sense to move this into rosbag2_cpp, it already has dependency on rosbag2_cpp?

Well, in theory, we can. It likely will give us the ability to reuse the same class in the `rosbag2_py::Player` and `rosbag2_py::Recorder` classes.
However, in this case we likely will need to make API/ABI breaking changes on the `rosbag2_cpp` layer.  
But we want to backport this PR down to the Humble if possible. And with the current design/approach it seems it will be doable since there are no API/ABI breaking changes so far.
Another point to consider is that moving this parameter's wrapper class to the `rosbag2_cpp` is too deep in the rosbag2 hierarchy and we unlikely ever will be using this newly created class on the `rosbag2_transport` or other really cpp layers. It is only useful on the `rosbag2_py` layer.
BTW, with the current `Arguments` class declaration inside `_transport.cpp`, nothing stops us from using it for the `rosbag2_py::Recorder` class as well.

File: rosbag2_py/src/rosbag2_py/_transport.cpp
Comment (r7vme) [reply rosbag2_py/src/rosbag2_py/_transport.cpp]: done

**Comment by r7vme** [comment]
@MichaelOrlov @fujitatomoya thanks for review. Addressed comments and added test, ready for review

**Review by fujitatomoya — APPROVED** [review]
lgtm with green CI, @r7vme thanks for iterating!

**Comment by fujitatomoya** [comment]
@r7vme can you rebase this there are several commits ahead in mainline rolling? and then i can start the CI.

File: rosbag2_py/src/rosbag2_py/_transport.cpp
Comment (r7vme) [reply rosbag2_py/src/rosbag2_py/_transport.cpp]: Left in the same file, but moved to unnamed namespace

**Comment by r7vme** [comment]
@fujitatomoya rebased

**Comment by fujitatomoya** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21010/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15314/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21691/)

**Review by MichaelOrlov — APPROVED** [review]
LGTM. Thanks!

**Comment by fujitatomoya** [comment]
@r7vme test failures are pretty much unrelated i think, but just in case can you also take a look?

**Comment by r7vme** [comment]
@fujitatomoya No tests related to rosbag failed as far as I can see. Most failures for aarch64 are `could not find definitions for vital function(s): 'abort' 'vfprintf'` and for Windows DDS compiler test failed `projectroot.src.core.ddsc.tests.CUnit_ddsc_whc_check_end_state`.

**Comment by fujitatomoya** [comment]
@MichaelOrlov i will leave this to you, since i do not have merge permission for this repo.

**Comment by MichaelOrlov** [comment]
@r7vme @fujitatomoya We have a breakage on CI currently. Will wait until https://github.com/ros2/ros2/pull/1560#pullrequestreview-2046562756 will be merged - then will re-run CI one more time.

**Comment by MichaelOrlov** [comment]
Pulls: ros2/rosbag2#1625
Gist: https://gist.githubusercontent.com/MichaelOrlov/9c08d8fab6ae11c1acdc4ac2e8992037/raw/ca588acacdb07218367a2e7b3b5a384acf202968/ros2.repos
BUILD args:  --packages-above-and-dependencies ros2bag rosbag2_py rosbag2_tests
TEST args:  --packages-above ros2bag rosbag2_py rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/13847
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21029/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15331/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21723/)

**Comment by MichaelOrlov** [comment]
@r7vme Could you please address one new warning https://ci.ros2.org/job/ci_windows/21723/msbuild/new/ on Windows CI build?

**Comment by MichaelOrlov** [comment]
@ros-pull-request-builder retest this please

**Comment by MichaelOrlov** [comment]
Re-running Windows CI after attempting to fix the warning.
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/21736/)

**Comment by MichaelOrlov** [comment]
@r7vme Could you please help to backport this PR to the Iron and Humble?
We will do an "automatic" backport to Jazzy via mergify when #1632 will land on it. 
However, backporting to the Iron and Humble via Mergefy doesn't make sense since the code base in the relevant places is different. It will be easier to do it manually.

**Comment by r7vme** [comment]
> Could you please help to backport this PR to the Iron and Humble?

@MichaelOrlov https://github.com/ros2/rosbag2/pull/1655 https://github.com/ros2/rosbag2/pull/1654

**Comment by MichaelOrlov** [comment]
https://github.com/Mergifyio backport jazzy

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-kilted-kaiju-release/43902/1

**Pull request merged**
